simulators/ethereum/pyspec/runner.go

```diff
@@ -0,0 +1,196 @@
+		var fixtureTests map[string] fixtureTest
+		if err := common.LoadJSON(path, &fixtureTests); err != nil {
+			t.Logf("invalid test file: %v, unable to load json", err)
+			return nil
+		}
+
+		// create testcase structure from fixtureTests
+		for name, fixture := range fixtureTests {
+			network := fixture.json.Network
+			// skip networks post merge
+			for _, skip := range []string{"Istanbul", "Berlin", "London"} {
+				if strings.Contains(network, skip) {
+					continue
+				}
+			}
+			// check network is valid 
+			if _, exist := envForks[network]; !exist {
+				return fmt.Errorf("network `%v` not defined in fork ruleset", network)
+			}
+
+			// extract genesis fields from fixture test
+			genesis := extractGenesis(fixture.json)
+
+			// extract payloads from each block
+			payloads := []*api.ExecutableData{}
+			for _, block := range fixture.json.Blocks {
+				gblock, _ := block.decodeBlock()
+				payload := api.BlockToExecutableData(gblock, common.Big0).ExecutionPayload
+				payloads = append(payloads, payload)
+			}
+
+			// extract post account information
+			postAlloc := fixture.json.Post
+
+			// define testcase struct adding the extracted fields
+			tc := testcase {
+				fixture: fixture, 
+				name: name, 
+				filepath: path,
+				genesis: genesis,  
+				payloads: payloads,
+				postAlloc: postAlloc,
+			}
+
+			// feed testcase to single worker within fixtureRunner()
+			fn(tc)
+		}
+		return nil
+
+	})
+}
+
+// run launches the client and runs the test case against it.
+func (tc *testcase) run(t *hivesim.T) {
+	start := time.Now()
+
+	t.Log("HIVE LOG --> Setting variables required for starting client.")
+	engineAPI := hive_rpc.HiveRPCEngineStarter{
+		ClientType: tc.clientType,
+		EnginePort: globals.EnginePortHTTP,
+		EthPort:    globals.EthPortHTTP,
+		JWTSecret:  globals.DefaultJwtTokenSecretBytes,
+	}
+	ctx := context.Background()
+	env := hivesim.Params{
+		"HIVE_FORK_DAO_VOTE": "1",
+		"HIVE_CHAIN_ID":      "1",
+		"HIVE_SKIP_POW":      "1",
+		"HIVE_NODETYPE": 	  "full",
+	}
+	tc.updateEnv(env)
+	t0 := time.Now()
+
+	// start client (also creates an engine RPC client internally)
+	t.Log("HIVE LOG --> Starting client with Engine API.")
+	engineClient, err := engineAPI.StartClient(t, ctx, tc.genesis, env, nil)
+	if err != nil {
+		t.Fatalf("HIVE FATAL --> Can't start client with Engine API: %v", err)
+	}
+	t1 := time.Now()
+
+	// send payloads
+	for blockNumber, payload :=  range tc.payloads {
+		_, err := engineClient.NewPayloadV2(context.Background(), payload)
+		if err != nil {
+			t.Fatalf("HIVE FATAL ---> Unable to send payload[%v] in test %s: %v ", blockNumber, tc.name, err)
+		}
+	}
+	t2 := time.Now()
+
+	// update head of "beacon" chain with the latest valid response
+	latestBlockHash := *engineClient.LatestNewPayloadResponse().LatestValidHash
+	fcState := &api.ForkchoiceStateV1{HeadBlockHash: latestBlockHash}
+	engineClient.ForkchoiceUpdatedV2(ctx, fcState, nil)
```

Comment:
We should check the result here, for `VALID` or `INVALID`, depending on whether the test contains a valid or invalid block.
And also check for the `err` in the response, in case the client did not respond correctly.

Reply:
It seems like the `err` in the response: `plStatus.ValidationError` is returning `null` for any invalid response.
```
{"jsonrpc":"2.0","id":1,"result":{"status":"INVALID","latestValidHash":null,"validationError":null}
```

I have committed the following nonetheless. 

```go
// send payloads and check response
latestValidHash := common.Hash{}
for blockNumber, payload := range tc.payloads {
	plException := tc.fixture.json.Blocks[blockNumber].Exception
	expectedStatus := "VALID"
	if plException != "" {
		expectedStatus = "INVALID"
	}
	// execute fixture block payload
	plStatus, plErr := engineClient.NewPayloadV2(context.Background(), payload)
	if plErr != nil {
		t.Fatalf("HIVE FATAL ---> ...`)
	}
	// update latest valid block hash
	if plStatus.Status == "VALID" {
		latestValidHash = *plStatus.LatestValidHash
	}
	// check payload status is expected from fixture
	if expectedStatus != plStatus.Status {
		t.Errorf(`HIVE ERROR ---> ...`)
	}
	// check error message if invalid
	if expectedStatus == "INVALID" && *plStatus.ValidationError != plException {
		t.Errorf(`HIVE ERROR ---> ...`)
	}
}
```

Reply:
Yes, I think it's an issue of the lack of error unification, most clients print different error strings here but it would be very difficult for us to match a string for an expected error on every client.

For now we could simply check for `INVALID`.